libnetwork/driver_plugin.py

```diff
@@ -70,13 +101,154 @@ def make_json_error(ex):
+
+
+@app.route('/IpamDriver.ReleasePool', methods=['POST'])
+def release_pool():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleasePool JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    json_response = {}
+    app.logger.debug("ReleasePool response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.RequestAddress', methods=['POST'])
+def request_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("RequestAddress JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    address = json_data["Address"]
+
+    if not address:
+        app.logger.debug("Auto assigning IP from Calico pools")
+
+        # No address requested, so auto assign from our pools.
+        num_v4 = 1 if pool_id == POOL_ID_V4 else 0
+        num_v6 = 1 if pool_id == POOL_ID_V6 else 0
+
+        # Auto assign an IP based on whether the IPv4 or IPv6 pool was selected.
+        # We auto-assign from all available pools with affinity based on our
+        # host.
+        ips_v4, ips_v6 = client.auto_assign_ips(num_v4, num_v6, None, None,
+                                                pool=(None, None),
+                                                hostname=hostname)
+        ips = ips_v4 + ips_v6
+        if not ips:
+            error_message = "There are no available IP addresses in the " \
+                            "configured Calico IP pools"
+            app.logger.error(error_message)
+            raise Exception(error_message)
+    else:
+        app.logger.debug("Reserving a specific address in Calico pools")
+        try:
+            ip_address = IPAddress(address)
+            rc = client.assign_ip(ip_address, None, {}, hostname=hostname)
+            if not rc:
+                error_message = "The address %s is already in " \
+                                "use" % str(ip_address)
+                app.logger.error(error_message)
+                raise Exception(error_message)
+            ips = [ip_address]
+        except PoolNotFound:
+            error_message = "The address %s is not in one of the configured " \
+                            "Calico IP pools" % str(ip_address)
+            app.logger.error(error_message)
+            raise Exception(error_message)
+
+    # We should only have one IP address assigned at this point.
+    assert len(ips) == 1, "Unexpected number of assigned IP addresses"
+
+    # Return the IP as a CIDR.
+    json_response = {
+        "Address": str(IPNetwork(ips[0])),
+        "Data": {}
+    }
+    app.logger.debug("RequestAddress response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.ReleaseAddress', methods=['POST'])
+def release_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleaseAddress JSON=%s", json_data)
+    address = json_data["Address"]
+
+    # Unassign the address.
+    client.release_ips({IPAddress(address)})
```

Comment:
Can you expand the comment to reassure the reader that it's OK if the release_ips fails. What happens if the release_ips throws an exception and will it generate a nice error for the user that's removing their container.